File: 3022-minimize-or-of-remaining-elements-using-operations/3022-minimize-or-of-remaining-elements-using-operations.java
class Solution {
    public int minOrAfterOperations(int[] nums, int k) {

        int answer = 0;
        int mask = 0;

        for (int bit = 29; bit >= 0; bit--) {

            int candidate = mask | (1 << bit);

            int groups = 0;
            int current = -1;

            for (int num : nums) {

                current &= num;

                if ((current & candidate) == 0) {
                    groups++;
                    current = -1;
                }
            }

            if (nums.length - groups <= k) {
                mask = candidate;
            } else {
                answer |= 1 << bit;
            }
        }

        return answer;
    }
}

File: 3022-minimize-or-of-remaining-elements-using-operations/3022-minimize-or-of-remaining-elements-using-operations.cpp
class Solution {
public:
    int minOrAfterOperations(vector<int>& nums, int k) {
        int res = 0, mask = 0;

        for (int i = 29; i >= 0; i--) {
            int next = mask | 1 << i;
            int sum = 0, cur = next;

            for (auto& c : nums) {
                cur &= c;
                sum += !cur;
                if (!cur) cur = next;
            }

            if (nums.size() - sum <= k) mask = next;
            else res |= 1 << i;
        }

        return res;
    }
};